jaxopt/_src/anderson.py

```diff
@@ -0,0 +1,197 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Anderson Acceleration for finding a fixed point in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+
+from dataclasses import dataclass
+
+import jax.numpy as jnp
+from jax.tree_util import tree_leaves, tree_structure
+
+from jaxopt._src import base
+from jaxopt._src import linear_solve
+from jaxopt._src.tree_util import tree_collapse, tree_uncollapse, tree_l2_norm, tree_sub
+
+
+class AndersonState(NamedTuple):
+  """Named tuple containing state information.
+  
+  Attributes:
+    iter_num: iteration number
+    value: pytree of current estimate of fixed point
+    error: residuals of current estimate
+    v_history: history of previous iterates
+    f_history: fixed_point_fun(v_history)
+  """
+  iter_num: int
+  value: Any
+  error: float
+  v_history: Any
+  f_history: Any
+
+
+@dataclass
+class AndersonAcceleration(base.IterativeSolver):
+  """Multi-dimensional fixed point solver using Anderson.
+
+  Read [1] to see the hypothesis ``fixed_point_fun`` must fulfil
+  to ensure convergence. In particular, if Banach fixed point theorem
+  hypothesis hold, Anderson acceleration will converge.
+  
+  [1] Pollock, Sara, and Leo Rebholz.
+  "Anderson acceleration for contractive and noncontractive operators."
+  arXiv preprint arXiv:1909.04638 (2019).
+
+  Attributes:
+    fixed_point_fun: a function ``fixed_point_fun(x, *args, **kwargs)``
+      returning a pytree with the same structure and type as x
+      each leaf must be an array (not a scalar)
+    history_size: size of history. Affect memory cost.
+    beta: momentum in Anderson updates. Default = 1.
+    maxiter: maximum number of iterations.
+    tol: tolerance (stoping criterion).
+    ridge: ridge regularization in solver.
+      Consider increasing this value if the solver returns ``NaN``.
+    has_aux: wether fixed_point_fun returns additional data. (default: False)
+      if True, the fixed-point is computed only with respect to first element of the sequence
+      returned. ``AndersonState.value`` will contain the full sequence returned.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto")
+  """
+  fixed_point_fun: Callable = None
+  history_size: int = 5
+  beta: float = 1.
+  maxiter: int = 100
+  tol: float = 1e-5
+  ridge: float = 1e-5
+  has_aux: bool = False
+  verbose: bool = False
+  implicit_diff: bool = False
```

Comment:
Does `implicit_diff=False` work?

Reply:
The tests work with both `implicit_diff=True` and `implicit_diff=False`.